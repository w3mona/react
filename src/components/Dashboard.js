import React from 'react';
import ListExpenses from './List';


const DashBoard = () =>(
    <div>
        <h2>Home</h2>
        <ListExpenses/>
    </div>
);
export default DashBoard;