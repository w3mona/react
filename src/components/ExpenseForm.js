import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';

import {removeExpense} from '../actions/expenses';




export default class ExpenseForm extends React.Component{
     //now = moment();
        //console.log(now.format('MMM Do, YYYY'));
        constructor(props){
        
            super(props);
            //console.log(props.expenses.createdAt,"vvv");
            this.state={
                
                id:props.expenses ? props.expenses.id: '',
                description:props.expenses ? props.expenses.description : '' ,
                notes:props.expenses ? props.expenses.notes : '',
                amount: props.expenses ? props.expenses.amount.toString() : '',
                createdAt: props.expenses ? moment(props.expenses.createdAt)   : moment(),
                calendarFocused:false,
                error: ''
            }

           // console.log(props.expenses);
        }

    desciptionChanged=(e)=>{
        const description = e.target.value;
        this.setState(()=>({description}));
    }

    notesChanged=(e)=>{
        const notes = e.target.value;
        //e.persist();
        this.setState(()=>({notes}));
    }

    amountChanged=(e)=>{
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.{1}\d{0,2})?$/)){
            this.setState(()=>({amount}));
        }
    }
    onDateChange=(createdAt)=>{
        if(createdAt){
            this.setState(()=>({createdAt}));
        }

        

    }

    onSubmitxx=(e)=>{
            e.preventDefault();
            if(!this.state.description || !this.state.amount){
                const error= "Error: Description or amount can't be empty";
                this.setState(()=>({error}));
                
            }
            else
            {
                this.setState(()=>({error:''})); 
            
              
                this.props.onSubmitxx({

                    description: this.state.description,
                    notes: this.state.notes,
                    amount: parseFloat(this.state.amount,10),
                    createdAt: this.state.createdAt.valueOf()
                })
            }
    }

    onFocusChange=({focused})=>{
        //console.log({focused});
        this.setState(()=>({calendarFocused: focused}));
    }
        render(){
            return(
                <div>
                {this.state.error && <p>{this.state.error}</p>}
                    <form onSubmit={this.onSubmitxx}>
                        <input type="text" placeholder="Description" autoFocus value={this.state.description} onChange={this.desciptionChanged}/>
                        <input type="number" placeholder="Amount" value={this.state.amount} onChange={this.amountChanged} />
                        <SingleDatePicker
                                date={this.state.createdAt} 
                                onDateChange={this.onDateChange} 
                                focused={this.state.calendarFocused} 
                                onFocusChange={this.onFocusChange } 
                                numberOfMonths={1}
                                isOutsideRange={()=>false}/>
                        <textarea placeholder="Add Notes To Expense" onChange={this.notesChanged}   value={this.state.notes} ></textarea>
                        <button>{this.state.id ? "Update Expense" : "Add Expense"}</button>

                        

                        
                    </form>
                </div>
            );
        }

}