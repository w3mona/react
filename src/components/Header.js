import React from 'react';
import {NavLink} from 'react-router-dom';

const Header=()=>(
  <header>
      <h1>Expensify</h1>
      <NavLink to="/" activeClassName="isActive"  exact={true} >Home</NavLink> 
      <NavLink to="/create" activeClassName="isActive" >Create</NavLink> 

      <NavLink to="/help" activeClassName="isActive">Help</NavLink> 
  </header>
);


export default Header;