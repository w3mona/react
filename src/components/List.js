import React from 'react';
import {connect} from 'react-redux';
import ListExpensesItem from './ListItem';
import getVisibleExpense from '../selectors/expense';
import ExpenseListFilter from './ListFilter';


//<p>Filters:{props.filters.text}</p>

export const ListExpenses = (props) =>(
    <div>
    <ExpenseListFilter/>
    {
        props.expenses.length===0 ?( "No Expenses" ): (
        
             props.expenses.map((expense )=>(
                <ListExpensesItem key={expense.id} {...expense} />
            ))
    )
        
    }
    
      
    </div>);


const mapStatesToProps=(state)=>{
    //return getVisibleExpense(state.expenses,state.filters)
    return{
        expenses:getVisibleExpense(state.expenses,state.filters) ,
        filters:state.filters
    }
}

export default connect(mapStatesToProps)(ListExpenses);