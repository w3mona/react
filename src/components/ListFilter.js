import React from 'react';
import {connect} from 'react-redux';
import {filterSerach,sortByDate,sortByAmount,setStartDate,setEndDate} from '../actions/filters';
import {DateRangePicker} from 'react-dates';

export class ExpenseListFilter extends React.Component{
    constructor(props){
            super(props);
            this.state={
                calendarFocused:null
            }
    }

    onDatesChange=({startDate,endDate})=>{
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);

    }



      filterTextChanged=(e)=>{
        this.props.SetText(e.target.value);
   
        }

        filterSelectChanged =(e)=>{
            
            if(e.target.value=="date")
            {
                
                this.props.SetByDate();
            }
            else if(e.target.value=="amount"){
                this.props.SetByAmount();
            }
           
        }
       


    onFocusChange=(calendarFocused)=>{
        this.setState(()=>({calendarFocused}));
    }

    render(){
        return (<div>
            <input 
            type="text" 
            value={this.props.filters.text} 
            onChange={this.filterTextChanged} 
            />

            <select onChange={this.filterSelectChanged}>
                <option value="date">Date</option>
                <option value="amount">Amount</option>
            </select>
            <DateRangePicker
            startDate={this.props.filters.startDate} 
            endDate={this.props.filters.endDate} 
            onDatesChange={this.onDatesChange} 
            focusedInput={this.state.calendarFocused} 
            onFocusChange={this.onFocusChange}  
            isOutsideRange={()=>false}
            numberOfMonths={1}
            showClearDates={true}
            ></DateRangePicker>
        </div>);
    }
}

const mapStateToProps=(state)=>({
    filters: state.filters
});

const mapDispatchToProps=(dispatch)=>({
    setStartDate: (startDate)=>dispatch(setStartDate(startDate)),
    setEndDate:(endDate)=> dispatch(setEndDate(endDate)),
    SetText:(text)=>dispatch(filterSerach({text})),
    SetByDate: ()=>dispatch(sortByDate()),
    SetByAmount:()=>dispatch(sortByAmount())
    
    

})



export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilter);

