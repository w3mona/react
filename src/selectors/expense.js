import moment from 'moment';


const getVisibleExpense=(expenses,{text,sortBy,startDate, endDate})=>{
    
    
    
    return expenses.filter((expense)=>{
        const createdAtMoment =moment(expense.createdAt);
        //console.log(moment(expense.createdAt));

        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment,'day') : true;
        const endDateMatch =  endDate ? endDate.isSameOrAfter(createdAtMoment,'day') : true;
        //const txtMatch = text.toLowerCase()===expense.description.toLowerCase();
        const txtMatch = expense.description.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && txtMatch;

    }).sort((a,b)=>{
        if(sortBy==="date"){
            return a.createdAt<b.createdAt ? 1:-1;
        }
        if(sortBy==="amount"){
            return a.amount<b.amount? 1:-1; 
        }
    });

}

export default getVisibleExpense;