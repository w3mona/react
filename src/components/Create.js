import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import {addExpense} from '../actions/expenses';

export class Create extends  React.Component{
  onSubmit=(expense)=>{
    this.props.onSubmityy(expense);
    this.props.history.push('/');
  }

  render(){
    return(
      <div>
      <h2>Add new Expense</h2>
      <ExpenseForm onSubmitxx={this.onSubmit}/>
    </div>
    );
    }

}


const mapDispatchToProps=(dispatch)=>(
    {
      onSubmityy:(expense)=>dispatch(addExpense(expense))
    }
  );
export default connect(undefined,mapDispatchToProps)(Create);