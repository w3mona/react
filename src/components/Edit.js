import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense,removeExpense} from '../actions/expenses';


export class EditClass extends React.Component{

    onSubmit=(expense)=>{
        this.props.onUpdate(expense,this.props.match.params.id );
        this.props.history.push("/");
    }

    delete=(e)=>{
        this.props.onDelete(this.props.match.params.id );
        this.props.history.push("/");
    }

    render(){
        return (<div>
            <h2>Edit param is {this.props.match.params.id} </h2>
            <ExpenseForm expenses={this.props.expenses}
            onSubmitxx={this.onSubmit}  />
             {this.props.match.params.id ? <button onClick={this.delete}>Delete</button> : ""}
          </div>);
    }
}



const mapStatesToProps =(state,props)=>{
             return {expenses: state.expenses.find((expense)=> expense.id === props.match.params.id)}
    }

    const mapDisptchToProps =(dispatch)=>({
        onUpdate:(expense,id)=>dispatch(editExpense({id:id,...expense})),
        onDelete:(id)=> dispatch(removeExpense({id}))
    });

      


    export default connect(mapStatesToProps,mapDisptchToProps)(EditClass);