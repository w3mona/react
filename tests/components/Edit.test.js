import React from 'react';
import {shallow} from 'enzyme';
import  {EditClass} from '../../src/components/Edit';
import expense from '../fixtures/expenses';

const match = {
    params : { 
                id : expense[0].id
             }
   }
let onUpdate,onDelete,wrapper,history;

beforeEach(()=>{
    onUpdate = jest.fn();
    onDelete = jest.fn();
    history= { push: jest.fn() };
    wrapper= shallow(<EditClass history={history} 
                            expenses={expense[0]} 
                            onUpdate={onUpdate} 
                            onDelete={onDelete} match={match} />);

});

test("Edit component render",()=>{
    expect(wrapper).toMatchSnapshot();
});

test("Edit render onupdate",()=>{


          wrapper.find('ExpenseForm').prop('onSubmitxx')(expense[0]);
        expect(history.push).toHaveBeenLastCalledWith('/');
        expect(onUpdate).toHaveBeenLastCalledWith(expense[0], expense[0].id);    
});

test("Edit render ondelete",()=>{

    wrapper.find("button").simulate("click");
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(onDelete).toHaveBeenLastCalledWith(expense[0].id);
   
});