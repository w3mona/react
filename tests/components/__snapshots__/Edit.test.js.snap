// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit component render 1`] = `
<div>
  <h2>
    Edit param is 
    1
     
  </h2>
  <ExpenseForm
    expenses={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "Gum",
        "id": "1",
        "note": "",
      }
    }
    onSubmitxx={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
